From 7d8c4054f461e575aab5ebf74eccf82825818d56 Mon Sep 17 00:00:00 2001
From: Alexey Kodanev <aleksei.kodanev@bell-sw.com>
Date: Mon, 6 Sep 2021 15:01:20 +0300
Subject: [PATCH] Makefile: link glibc-string lib

libglibc-string.a is unpacked and all it's object files are
embedded/linked into the final libc.a/libc.so files.

In addition, provide a separate static library called
libc-noglibc-string.a with no libglibc-string.a objects.
---
 Makefile | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index e8cc4436..8ab40e94 100644
--- a/Makefile
+++ b/Makefile
@@ -64,7 +64,7 @@ ALL_INCLUDES = $(sort $(INCLUDES:$(srcdir)/%=%) $(GENH:obj/%=%) $(ARCH_INCLUDES:
 EMPTY_LIB_NAMES = m rt pthread crypt util xnet resolv dl
 EMPTY_LIBS = $(EMPTY_LIB_NAMES:%=lib/lib%.a)
 CRT_LIBS = $(addprefix lib/,$(notdir $(CRT_OBJS)))
-STATIC_LIBS = lib/libc.a
+STATIC_LIBS = lib/libc.a lib/libc-noglibc-string.a
 SHARED_LIBS = lib/libc.so
 TOOL_LIBS = lib/musl-gcc.specs
 ALL_LIBS = $(CRT_LIBS) $(STATIC_LIBS) $(SHARED_LIBS) $(EMPTY_LIBS) $(TOOL_LIBS)
@@ -158,11 +158,20 @@ obj/%.lo: $(srcdir)/%.S
 obj/%.lo: $(srcdir)/%.c $(GENH) $(IMPH)
 	$(CC_CMD)
 
-lib/libc.so: $(LOBJS) $(LDSO_OBJS)
+unpack_glibc_string:
+	test -d glibc-string-ar || mkdir glibc-string-ar && cd glibc-string-ar && $(AR) -x ../libglibc-string.a
+
+lib/libc.so: $(LOBJS) $(LDSO_OBJS) unpack_glibc_string
 	$(CC) $(CFLAGS_ALL) $(LDFLAGS_ALL) -nostdlib -shared \
-	-Wl,-e,_dlstart -o $@ $(LOBJS) $(LDSO_OBJS) $(LIBCC)
+	-Wl,-e,_dlstart -o $@ $(LOBJS) glibc-string-ar/* $(LDSO_OBJS) $(LIBCC)
+
+lib/libc.a: $(AOBJS) unpack_glibc_string
+	rm -f $@
+	$(AR) rc $@ $(AOBJS) glibc-string-ar/*
+	$(RANLIB) $@
 
-lib/libc.a: $(AOBJS)
+# Create static library with no glibc-string code, just musl
+lib/libc-noglibc-string.a: $(AOBJS)
 	rm -f $@
 	$(AR) rc $@ $(AOBJS)
 	$(RANLIB) $@
-- 
2.25.1

